Refactoring to improve the code style.

matejak opened the pull request:
A harmless refactor that should increase maintainability.

pep8speaks commented:
Hello @matejak! Thanks for submitting the PR.

 - In the file [`shared/misc/generate-contributors.py`](https://github.com/OpenSCAP/scap-security-guide/blob/49bf71672d8814214608e422678495153b6ab70d/shared/misc/generate-contributors.py), following are the PEP8 issues :

> [Line 13:1](https://github.com/OpenSCAP/scap-security-guide/blob/49bf71672d8814214608e422678495153b6ab70d/shared/misc/generate-contributors.py#L13): [E122](https://duckduckgo.com/?q=pep8%20E122) continuation line missing indentation or outdented

mpreisler commented:
Great work!

I found this unexpected hunk:
```
-<contributor>John Hooks &lt;hooksie11@gmail.com&gt;</contributor>
+<contributor>John Hooks &lt;jhooks@starscream.pa.jhbcomputers.com&gt;</contributor>
```

I am not sure why it happens but some of the behavior changed.

mpreisler commented:
Also, consider me a cynic but this refactoring makes the code much longer. Some of the changes are great - defaultdict for example, but some other changes are IMO needless (extracting functions only used once bloats the code base). I don't feel strongly about this but IMO this goes too far.

matejak commented:
I believe that I have fixed the ordering issues (i.e. use other than the very last name), so Python 2 and 3 should be the same.
Concerning functions count, I have the opposite view - there are still functions that do more than one thing and some also have lots of execution paths, but I consider practical to stop extraction at this point.
I have also added unit tests and coverage support for the SSG build system.

matejak commented:
Jenkins check needs https://github.com/OpenSCAP/jenkins/pull/44 for merge.

Why pytest over doctests:
* Having tests in docstrings would increase the file size, making it more difficult to read and navigate.
* `pytest` provides many features that help to make test code tidy and also debugging easy if tests fail.
* `pytest` make it easy to support coverage reports.

mpreisler commented:
doctests also act as documentation though, which is awesome

mpreisler commented on cmake/SSGCommon.cmake:
```diff
@@ -226,6 +226,23 @@ if (SSG_SHELLCHECK_BASH_FIXES_VALIDATION_ENABLED AND SHELLCHECK_EXECUTABLE AND "
 endmacro()
 
+
+set(PYTEST_COVERAGE_OPTIONS)
```
what is this for? do you mean `unset(PYTEST_COVERAGE_OPTIONS)`?

mpreisler replied:
why is this variable guarded by ENABLE_PYTHON_COVERAGE but not the other stuff? If I set it to disabled it will all fail, right?

mpreisler commented:
Btw, please open a new PR next time. This PR is now doing 2 things that are unrelated, there is no way to reject one part and accept the other and it's messy.

mpreisler commented:
Hmm, odd, this is failing its own unit tests:
```
The following tests FAILED:
	  1 - python-unit-misc (Failed)
	  2 - python-unit-utils (Failed)
```

matejak commented:
Failing tests are expected, as Jenkins needs `pytest` installed for tests to work - see https://github.com/OpenSCAP/jenkins/pull/44

mpreisler commented:
> Failing tests are expected, as Jenkins needs pytest installed for tests to work - see OpenSCAP/jenkins#44

ok, makes sense, have you deployed the playbooks? then it should be installed even if it's not merged

mpreisler commented:
> Failing tests are expected, as Jenkins needs pytest installed for tests to work - see OpenSCAP/jenkins#44

Please check for pytest as part of configure and disable the tests if pytest is not available. I consider that a merge blocker.

matejak commented:
@mpreisler Regarding Jenkins tests/setup, I have the following plan:
* Test on Jenkins without pytest (the pytest dependency is now optional, so it should pass).
* Install pytest on Jenkins by applying playbooks.
* Let this PR be tested again, it should pass, but tests should also include the new unit tests.

matejak commented:
@openscap-ci test this please

mpreisler commented:
> @mpreisler Regarding Jenkins tests/setup, I have the following plan:
> 
> Test on Jenkins without pytest (the pytest dependency is now optional, so it should pass).
> Install pytest on Jenkins by applying playbooks.
> Let this PR be tested again, it should pass, but tests should also include the new unit tests.

Sounds good, pytest is now avaialble on all workers. I will test this once jenkins recovers.

mpreisler commented:
@openscap-ci test this please

mpreisler commented:
@openscap-ci add to whitelist

mpreisler commented:
@matejak this needs a rebase

matejak commented:
@mpreisler rebased.

matejak commented:
@openscap-ci test this please

mpreisler commented:
ACK
